pkg/serviceaccount/claims.go

```diff
@@ -199,6 +227,37 @@ func (v *validator) Validate(ctx context.Context, _ string, public *jwt.Claims,
 		podUID = podref.UID
 	}
 
+	var nodeName, nodeUID string
+	if noderef != nil {
+		switch {
+		case podref != nil && utilfeature.DefaultFeatureGate.Enabled(features.ServiceAccountTokenPodNodeInfo):
+			// for pod-bound tokens, just extract the node claims
+			nodeName = noderef.Name
+			nodeUID = noderef.UID
+		case podref == nil:
+			if !utilfeature.DefaultFeatureGate.Enabled(features.ServiceAccountTokenNodeBindingValidation) {
+				klog.V(4).Infof("ServiceAccount token is bound to a Node object, but the node bound token validation feature is disabled")
+				return nil, fmt.Errorf("token is bound to a Node object but the %s feature gate is disabled", features.ServiceAccountTokenNodeBindingValidation)
+			}
+
+			node, err := v.getter.GetNode(noderef.Name)
+			if err != nil {
+				klog.V(4).Infof("Could not retrieve node object %q for service account %s/%s: %v", noderef.Name, namespace, saref.Name, err)
+				return nil, errors.New("service account token has been invalidated")
+			}
+			if node.DeletionTimestamp != nil && node.DeletionTimestamp.Time.Before(invalidIfDeletedBefore) {
+				klog.V(4).Infof("Node %q is deleted and awaiting removal for service account %s/%s", node.Name, namespace, saref.Name)
+				return nil, errors.New("service account token has been invalidated")
+			}
+			if noderef.UID != string(node.UID) {
+				klog.V(4).Infof("Node UID no longer matches %s: %q != %q", noderef.Name, string(node.UID), noderef.UID)
+				return nil, fmt.Errorf("node UID (%s) does not match service account node ref claim (%s)", node.UID, noderef.UID)
+			}
```

Comment:
move this above the deletionTimestamp check (and do the same for pod above). if the uid doesn't match, the deletionTimestamp is irrelevant (and making that the returned error is confusing when the real issue is that we're dealing with a different instance)

Reply:
do you want me to file a separate PR to change this for bound pod & secret objects too, or include it here? This will change the error response message for these tokens, which i think we could consider a bugfix/an acceptable change in API behaviour?

Reply:
You can include it here. It only changes the message, not the response code or validity of the token, so it's an acceptable tweak